### ISSUE TITLE
bug: side effects with reassignment in ionic 6 + webpack, vite

### ISSUE THREAD

**Issue by ptmkenny** [ionic-team/ionic-framework#24280]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

- [ ] v4.x
- [ ] v5.x
- [X] v6.x

### Current Behavior

I have an Ionic React v5 app that I upgraded to v6. When I did so and inspected the bundle size:

`react-scripts build --stats; webpack-bundle-analyzer build/bundle-stats.json`

I noticed that the node-modules bundle is now 3x larger (from ~500 kB to 1.4 MB).

Here's my Ionic 5 app bundle:

<img width="1867" alt="ionic5" src="https://user-images.githubusercontent.com/1451472/143669504-d821e8ec-c906-44d7-9a66-d7d2129aa86d.PNG">
 
And here's the Ionic 6 bundle:

<img width="1873" alt="ionic6-rc3" src="https://user-images.githubusercontent.com/1451472/143669532-e05ad0b7-95d4-4074-b24f-404cecb9c060.PNG">

The only changes between the code for Ionic v5 and v6 are a few lines to update a datetime component, so these builds are basically identical except for the Ionic version.

### Expected Behavior

When upgrading to the next major Ionic version, the bundle size should be roughly the same or smaller than the previous version.

I read in the [beta announcement](https://ionicframework.com/blog/announcing-the-ionic-framework-v6-beta/) that even smaller bundles are coming to Vue first, and then React and Angular, but I didn't except a massive size increase.

### Steps to Reproduce

Update a moderately sized app from v5 to v6.

### Ionic Info

```
Ionic:

   Ionic CLI       : 6.18.0 (/home/me/.local/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/react 6.0.0-rc.3

Capacitor:

   Capacitor CLI      : 3.3.2
   @capacitor/android : 3.3.2
   @capacitor/core    : 3.3.2
   @capacitor/ios     : 3.3.2

Utility:

   cordova-res : 0.15.3
   native-run  : 1.5.0

System:

   NodeJS : v14.16.0 (/usr/local/bin/node)
   npm    : 7.19.1
   OS     : Linux 4.19
```

**Comment by sean-perkins** [comment]
@ptmkenny can you provide a sample repository reproducing this issue?

On an initial look, it doesn't appear the build output is a production build, where the treeshaking happens.

**Comment by sean-perkins** [comment]
We've created a few examples locally with Vue and React and have been able to identify a relationship between Stencil version upgrades and bundle size increases.

**Comment by liamdebeasi** [comment]
Thanks! We identified two issues that are causing treeshaking to not behave as expected.

1. The framework wrappers we use from Stencil had a sideeffect that caused all components to get pulled in. This was addressed in https://github.com/ionic-team/stencil-ds-output-targets/pull/208 and a fix will be in an upcoming release of Ionic 6.
2. There is a potential side effect caused by the `proxyCustomElement` function in Stencil that only affects Webpack 4 (Webpack 5 seems to work fine). This one requires a bit more investigation, but we are working closely with the Stencil team to fix this.

Treeshaking should be fixed for Webpack 5, Vite, Rollup, ES Build, etc users in the next release of Ionic 6. It should be slightly better for Webpack 4 users, but that `proxyCustomElement` issue will need to be resolved before treeshaking can work as intended. I will keep this open until that second issue is resolved. Thanks!

**Comment by piotr-cz** [comment]
@liamdebeasi 

> fix will be in an upcoming release of Ionic 6.

It's not included in the *6.0.0-rc.4*, isn't it?
Because there very small size difference after update (4.43 -> 4.38MB).

I've noticed significant bundle size increase after Swiper has been added in *6.0.0-rc.3* but this is probably a general problem with tree shaking when using Vite.

**Comment by liamdebeasi** [comment]
It is included in 6.0.0-rc.4. As I mentioned in https://github.com/ionic-team/ionic-framework/issues/24280#issuecomment-986951457, you should only notice a small improvement at the moment because of the lingering `proxyCustomElement` side effect inside of Stencil that we are still investigating. Interestingly, this only impacts Webpack as Vite/Rollup/ESBuild are not impacts, so I am wondering if this is a bug in Webpack.

**Comment by piotr-cz** [comment]
@liamdebeasi Thank you for explaining

**Comment by edosenseidotcom** [comment]
Thanks for the updates!

@liamdebeasi I updated to Ionic 6.0 official, and since you said Webpack 5 seems to be working, [based on the CRA 5 alpha discussion](https://github.com/facebook/create-react-app/discussions/11278), I updated my Ionic app to CRA5 alpha:

`    "react-scripts": "5.0.0-next.58+657739fb",`

My app builds, and I can use it, but the bundle size is still huge (about 3x what it used to be):

<img width="1900" alt="ionic6-cra5-bundle-size" src="https://user-images.githubusercontent.com/78798096/145522419-ae8fc357-ab98-4169-9c94-8ba95506fdd2.png">

This is building with CRA5 alpha, which I understand is now using Webpack 5 under the hood.

**Comment by ptmkenny** [comment]
Thanks for the updates!

@liamdebeasi I updated to Ionic 6.0 official, and since you said Webpack 5 seems to be working, [based on the CRA 5 alpha discussion](https://github.com/facebook/create-react-app/discussions/11278), I updated my Ionic app to CRA5 alpha:

`    "react-scripts": "5.0.0-next.58+657739fb",`

My app builds, and I can use it, but the bundle size is still huge (about 3x what it used to be):

<img width="1900" alt="ionic6-cra5-bundle-size" src="https://user-images.githubusercontent.com/1451472/145522581-4220def8-c9d1-4ae6-99c3-be927395f15b.png">

This is building with CRA5 alpha, which I understand is now using Webpack 5 under the hood.

**Comment by liamdebeasi** [comment]
Darn, seems like Webpack 5 is still impacted. Thanks for the additional info!

**Comment by liamdebeasi** [comment]
Following up on this: This issue is happening due to a bug in Webpack. I have reported the issue here: https://github.com/webpack/webpack/issues/14963

We are investigating potential workarounds and will follow up here when we have more to share.

**Comment by liamdebeasi** [comment]
The fix for this was merged into the Stencil codebase recently: https://github.com/ionic-team/stencil/commit/5dccc856cdaf19592d2a08f0f9b945b8bf2c4f7c

We are waiting on a stable Stencil release before updating Ionic.

**Comment by darrylnoakes** [comment]
I noticed the change between Ionic 6.0.2 and 6.0.8:

**Now**:

- Vite 2.8.4
- Ionic 6.0.8

```
✓ 242 modules transformed.
dist/index.html                         1.31 KiB
dist/assets/other.d5baf75a.js           0.24 KiB / gzip: 0.16 KiB
dist/assets/LoginPage.e8906c5f.js       5.26 KiB / gzip: 2.34 KiB
dist/assets/useIonicMode.749331d3.js    0.26 KiB / gzip: 0.20 KiB
dist/assets/SettingsPage.fd930851.js    3.51 KiB / gzip: 1.41 KiB
dist/assets/NotFoundPage.bb5b46d7.js    0.60 KiB / gzip: 0.35 KiB
dist/assets/swipe-back.bf0a77ce.js      0.55 KiB / gzip: 0.41 KiB
dist/assets/index.93ebd0c1.js           11.12 KiB / gzip: 4.34 KiB
dist/assets/tap-click.70c79e19.js       1.93 KiB / gzip: 0.93 KiB
dist/assets/SongView.69554a7b.js        14.09 KiB / gzip: 5.01 KiB
dist/assets/status-tap.55ffb348.js      0.49 KiB / gzip: 0.35 KiB
dist/assets/input-shims.ad083b7c.js     4.43 KiB / gzip: 1.89 KiB
dist/assets/SettingsPage.b0bcb306.css   2.07 KiB / gzip: 0.58 KiB
dist/assets/LoginPage.b6c175b6.css      4.55 KiB / gzip: 1.23 KiB
dist/assets/SongView.ab732b20.css       6.61 KiB / gzip: 1.36 KiB
dist/assets/index.278183c8.css          27.40 KiB / gzip: 5.16 KiB
dist/assets/swiper.bundle.a76d5250.js   93.85 KiB / gzip: 25.25 KiB
dist/assets/vendor.cef1e577.js          1017.79 KiB / gzip: 214.57 KiB
```

**"Update to stable"**:

- Vite 2.6.10
- Ionic 6.0.2

```
✓ 219 modules transformed.
dist/assets/music_note.9426a5a7.svg     0.28 KiB
dist/index.html                         1.32 KiB
dist/assets/other.bc05ace7.js           0.24 KiB / gzip: 0.15 KiB
dist/assets/LoginPage.9b789caf.js       4.17 KiB / gzip: 1.79 KiB
dist/assets/SettingsPage.3e4fceaa.js    3.22 KiB / gzip: 1.39 KiB
dist/assets/index.52b46503.js           10.18 KiB / gzip: 3.96 KiB
dist/assets/useMode.77daeedb.js         0.31 KiB / gzip: 0.21 KiB
dist/assets/NotFoundPage.5d12e648.js    0.62 KiB / gzip: 0.36 KiB
dist/assets/swipe-back.7db6ed39.js      0.55 KiB / gzip: 0.41 KiB
dist/assets/tap-click.93a07216.js       1.93 KiB / gzip: 0.93 KiB
dist/assets/SongView.54eb2248.js        10.26 KiB / gzip: 4.00 KiB
dist/assets/status-tap.d36ee3c7.js      0.49 KiB / gzip: 0.35 KiB
dist/assets/keyboard.39e4bf75.js        1.18 KiB / gzip: 0.59 KiB
dist/assets/input-shims.6b502cf4.js     4.43 KiB / gzip: 1.89 KiB
dist/assets/SettingsPage.29d87d36.css   2.19 KiB / gzip: 0.63 KiB
dist/assets/SongView.a2bb2425.css       1.51 KiB / gzip: 0.54 KiB
dist/assets/LoginPage.7f1c0244.css      4.26 KiB / gzip: 1.18 KiB
dist/assets/index.1d9bcac9.css          27.31 KiB / gzip: 5.11 KiB
dist/assets/vendor.418089c1.js          518.58 KiB / gzip: 130.21 KiB
```

**Comment by darrylnoakes** [comment]
Hmm...

> I noticed the change between Ionic 6.0.2 and 6.0.8.

It seems to be a change in _Vite_ the changed the sizes.
I tried with some different combos.

**Vite 2.6.10 + Ionic 6.0.2**

```
dist/assets/index.58199c5e.js           11.12 KiB / gzip: 4.34 KiB
dist/assets/index.ba26cccb.css          27.39 KiB / gzip: 5.16 KiB
dist/assets/vendor.f8c31942.js          547.73 KiB / gzip: 137.93 KiB
```

**Vite 2.6.10 + Ionic 6.0.8**

```
dist/assets/index.80267744.js           11.12 KiB / gzip: 4.35 KiB
dist/assets/index.278183c8.css          27.40 KiB / gzip: 5.16 KiB
dist/assets/vendor.cf3e1419.js          545.99 KiB / gzip: 137.97 KiB
```

**Vite 2.8.4 + Ionic 6.0.2**

```
dist/assets/index.4a8c4790.js           11.12 KiB / gzip: 4.35 KiB
dist/assets/index.ba26cccb.css          27.39 KiB / gzip: 5.16 KiB
dist/assets/swiper.bundle.6deadf64.js   93.85 KiB / gzip: 25.25 KiB
dist/assets/vendor.5d1d29a2.js          1022.79 KiB / gzip: 215.34 KiB
```

** Vite 2.8.4 + Ionic 6.0.8**

```
dist/assets/index.4a8c4790.js           11.12 KiB / gzip: 4.35 KiB
dist/assets/index.ba26cccb.css          27.39 KiB / gzip: 5.16 KiB
dist/assets/swiper.bundle.6deadf64.js   93.85 KiB / gzip: 25.25 KiB
dist/assets/vendor.5d1d29a2.js          1022.79 KiB / gzip: 215.34 KiB
```

**Comment by darrylnoakes** [comment]
The change happens in Vite between 2.6.14 (latest 2.6.x) and 2.7.0.

**Comment by darrylnoakes** [comment]
Relevant change is in Vite 2.7.0-beta.5.

Specifically, these changes in `packages/vite/src/node/plugins/resolve.ts`:

```diff
- // check for deep import, e.g. "my-lib/foo"
- const deepMatch = nestedPath.match(deepImportRE)
-
- const pkgId = deepMatch ? deepMatch[1] || deepMatch[2] : nestedPath
+ const possiblePkgIds: string[] = []
+  for (let prevSlashIndex = -1; ; ) {
+   let slashIndex = nestedPath.indexOf('/', prevSlashIndex + 1)
+   if (slashIndex < 0) {
+     slashIndex = nestedPath.length
+   }
+
+   const part = nestedPath.slice(
+     prevSlashIndex + 1,
+     (prevSlashIndex = slashIndex)
+   )
+   if (!part) {
+      break
+    }
+
+   // Assume path parts with an extension are not package roots, except for the
+   // first path part (since periods are sadly allowed in package names).
+   // At the same time, skip the first path part if it begins with "@"
+   // (since "@foo/bar" should be treated as the top-level path).
+  if (possiblePkgIds.length ? path.extname(part) : part[0] === '@') {
+     continue
+  }
+
+   const possiblePkgId = nestedPath.slice(0, slashIndex)
+   possiblePkgIds.push(possiblePkgId)
+ }
```

```diff
- const pkg = resolvePackageData(pkgId, basedir, options.preserveSymlinks)
+ const pkgId = possiblePkgIds.reverse().find((pkgId) => {
+   pkg = resolvePackageData(pkgId, basedir, options.preserveSymlinks)
+   return pkg
+ })
```

Removing that `reverse()` (see second diff) fixes it...
Causes these changes:

```
normal -> reversed

@ionic/core ->   @ionic/core/components
@stencil/core -> @stencil/core/internal/app-data
@stencil/core -> @stencil/core/internal/client
ionicons ->      ionicons/components
ionicons ->      ionicons/icons
```

**Comment by darrylnoakes** [comment]
Via some more modding of Vite's guts to choose a hardcoded package name for these cases, I've narrowed it down to the change between `@ionic/core` and `@ionic/components/core`.

**Comment by liamdebeasi** [comment]
Thanks for the issue. This has been resolved via https://github.com/ionic-team/ionic-framework/pull/24895, and a fix will be available in an upcoming release of Ionic Framework.

There are two issues being discussed in this thread:

1. There was a bug in Stencil where bundle sizes were larger than normal due to an extra reassignment. The linked PR fixes this issue in Ionic by updating our Stencil dependency to 2.14.1. This should result in a noticeable reduction in bundle size for both Vite and Webpack users.
2. There is a change in Vite that is causing Swiper.js + `ion-slides` to get bundled with applications, even if they are not being used. This was noted in https://github.com/ionic-team/ionic-framework/issues/24280#issuecomment-1054695925.

The linked PR does not do anything to address issue 2 as this is due to a change in Vite, not Ionic. The way we bundle Swiper.js in Ionic is a bit hacky, so I am not surprised this is getting incorrectly pulled in. The good news is that this Swiper code is being removed in the next major release of Ionic, so this issue should go away on its own. For now, you may want to consider staying with Vite 2.6.x. Alternatively, you could file an issue on the [Vite repo](https://github.com/vitejs/vite) to see if the Vite team is willing to work on a fix.

I am going to close this issue. Thanks!

**Comment by liamdebeasi** [comment]
Thanks for the issue. This has been resolved via https://github.com/ionic-team/ionic-framework/pull/24895, and a fix will be available in an upcoming release of Ionic Framework.

There are two issues being discussed in this thread:

1. There was a bug in Stencil where bundle sizes were larger than normal due to an extra reassignment. The linked PR fixes this issue in Ionic by updating our Stencil dependency to 2.14.1. This should result in a noticeable reduction in bundle size for both Vite and Webpack users.
2. There is a change in Vite that is causing Swiper.js + `ion-slides` to get bundled with applications, even if they are not being used. This was noted in https://github.com/ionic-team/ionic-framework/issues/24280#issuecomment-1054695925.

The linked PR does not do anything to address issue 2 as this is due to a change in Vite, not Ionic. The way we bundle Swiper.js in Ionic is a bit hacky, so I am not surprised this is getting incorrectly pulled in. The good news is that this Swiper code is being removed in the next major release of Ionic, so this issue should go away on its own. For now, you may want to consider staying with Vite 2.6.x. Alternatively, you could file an issue on the [Vite repo](https://github.com/vitejs/vite) to see if the Vite team is willing to work on a fix.

I am going to close this issue. Thanks!

**Comment by liamdebeasi** [comment]
Thanks for the issue. This has been resolved via https://github.com/ionic-team/ionic-framework/pull/24895, and a fix will be available in an upcoming release of Ionic Framework.

There are two issues being discussed in this thread:

1. There was a bug in Stencil where bundle sizes were larger than normal due to an extra reassignment. The linked PR fixes this issue in Ionic by updating our Stencil dependency to 2.14.1. This should result in a noticeable reduction in bundle size for both Vite and Webpack users.
2. There is a change in Vite that is causing Swiper.js + `ion-slides` to get bundled with applications, even if they are not being used. This was noted in https://github.com/ionic-team/ionic-framework/issues/24280#issuecomment-1054695925.

The linked PR does not do anything to address issue 2 as this is due to a change in Vite, not Ionic. The good news is that this Swiper code is being removed in the next major release of Ionic, so this issue should go away on its own. For now, you may want to consider staying with Vite 2.6.x. Alternatively, you could file an issue on the [Vite repo](https://github.com/vitejs/vite) to see if the Vite team is willing to work on a fix.

I am going to close this issue. Thanks!
